bug fix and adding recent updates to vignettes

cristinamullin opened the pull request:
- [x] Check vignettes to make sure recent updates have been incorporated
- [x] Consider adding tests to avoid issues in the future

cristinamullin committed:
add new TADA_ConvertSpecialChars clean = T option

cristinamullin committed:
fix TN and TP harmonization

add unit to grouping cols for TADA_AggregateMeasurements, fix TN and TP bug, add combos to harmonization table, update vignette

cristinamullin committed:
Update TADAModule1.Rmd

remove intermediate variables and update scatterplot examples

cristinamullin committed:
Update TADAModule1.Rmd

cristinamullin committed:
removes TADA_AutoFilter and adds functionality to TADA_RetainRequired

Combine TADA_AutoFilter with TADA_RetainRequired and delete TADA_AutoFilter

cristinamullin committed:
update vignettes

incorporate changes made to TADA_FindQCActivities, TADA_RetainRequired, and TADA_AutoFilter

cristinamullin committed:
Update TADA_RetainRequired.Rd

cristinamullin committed:
Update TADAModule1.Rmd

cristinamullin committed:
update refs

cristinamullin committed:
update example data

hillarymarler commented:
> Need to look into these tomorrow:
> 
> ```r
> ── Failed tests ─────────────────────────────────────────────────────────────────────────────
> Failure (test-TADARefTables.R:28:3): No combos were missed in NP key from harmonization table
> dim(orphs)[1] == 0 is not TRUE
> 
> `actual`:   FALSE
> `expected`: TRUE 
> 
> Failure (test-Transformations.R:21:3): np summation key matches nutrient harmonization ref
> any(is.na(check$np)) is not FALSE
> 
> `actual`:   TRUE 
> `expected`: FALSE
> ```

I took a look at this and it looks like the following may be missing from TADA_GetNutrientSummationRef

<img width="422" height="59" alt="image" src="https://github.com/user-attachments/assets/af1a5d6e-1155-46d4-a3aa-4fd32851f1e7" />

cristinamullin commented:
#588 I will work to incorporate this feedback into this PR. Also, thanks @hillarymarler for looking into the missing combinations. They will need to get prioritized for inclusion in the TN/TP calculations in that ref table. I have not edited this since Elise and I worked on it when she was an ORISE with our branch. I'll take a look & may loop the TADA team in to review the logic.

cristinamullin committed:
add missing rows to NP/TN summation key, add metadata to output

cristinamullin committed:
Update RequiredCols.R

cristinamullin commented:
I found a bug in TADA_ConvertSpecialChars where we are getting new columns with TADA.TADA. if the function input column already contains the TADA prefix. Writing a test to double check for this issue.

```r
test_that("Column names do not contain the pattern 'TADA.TADA.'", {
  test_TADA.TADA. <-
    TADA_ConvertSpecialChars(
      Data_Nutrients_UT,
      "TADA.DetectionQuantitationLimitMeasure.MeasureValue"
    )
  # Create a logical vector indicating which columns contain the pattern
  pattern_found <- grepl("TADA.TADA.", colnames(test_TADA.TADA.))
  
  # Test should pass if none of the columns contain the pattern
  expect_false(any(pattern_found), info = "Some column names contain the pattern 'TADA.TADA.'")
})
```

cristinamullin committed:
add tests

cristinamullin committed:
only run if not run previously

cristinamullin committed:
define flag col and add tests

add tests and examples for TADA_ConvertSpecialChars and TADA_CalculateTotalNP

wokenny13 commented:
I am looking at the example in the Module 1 vignette and just for phosphorus ( TADAProfileClean5)

It seems like there are only 3 other phosphorus TADA.ComparableDataIdentifier in our data frame. Since they are already as P would it then be converted as is to Total Phosphorus? Should we expect the Total Phosphorus, Mixed Forms to only increase by a maximum of (the sum of 62 + 8 + 125) 195 rows once you run TADA_CalculateTotalNP?

It seems like it goes from 581 rows to 1180? Are duplicate rows being added on to the Total Phosphorus, Mixed Forms summation?

<img width="625" height="119" alt="image" src="https://github.com/user-attachments/assets/e7fa6353-08fd-4508-9372-229bf7eea520" />

<img width="501" height="79" alt="image" src="https://github.com/user-attachments/assets/f790c70e-227d-4511-bfd1-e2fe5c760500" />

cristinamullin commented:
The TN/TP function is not working correctly... 

```r
today <- Sys.Date()
oneyearago <- as.character(today - 1 * 365)
testdat <- TADA_DataRetrieval(statecode = "UT", 
                              startDate = oneyearago, 
                              characteristicType = "Nutrient",
                              ask = FALSE)

testdat <- TADA_SimpleCensoredMethods(testdat,
                                      nd_method = "multiplier",
                                      nd_multiplier = 0.5,
                                      od_method = "as-is",
                                      od_multiplier = "null")

testdat2 <- TADA_CalculateTotalNP(testdat, daily_agg = "max")

unique(testdat2$TADA.ResultMeasureValueDataTypes.Flag)

# Filter data to show only rows where TADA.ResultMeasureValue is NA
filtered_data <- testdat2[is.na(testdat$TADA.ResultMeasureValue), ]

unique(filtered_data$TADA.NutrientSummation.Flag)
unique(filtered_data$TADA.ResultMeasureValueDataTypes.Flag)

# isolate rows that are NA related to TP and TN function
filtered_data_2 <- filtered_data[!(
  filtered_data$TADA.ResultMeasureValueDataTypes.Flag %in% 
    c("TP estimated from one or more subspecies.", 
      "TN estimated from one or more subspecies.")
)]

# isolate rows that are NA related to TP and TN function
# reduces number of columns... why
filtered_data_3 <- filtered_data[!(
  filtered_data$TADA.NutrientSummation.Flag %in% 
    c("Used to calculate Total Phosphorus as P.", 
      "Used to calculate Total Nitrogen as N." #,
      #"Not used to calculate Total N or P."
      )
)]

# these should not be used to calculate TN and TP if result value is NA
# and TADA.ResultMeasureValueDataTypes.Flag equals any of the following...
# [1] "NA - Not Available"                                        
# [2] "Result Value/Unit Cannot Be Estimated From Detection Limit"
# [3] "Text"
```

cristinamullin committed:
revert column additions

wokenny13 commented:
Yes, I can take a look into this

cristinamullin commented on R/RequiredCols.R:
```diff
@@ -449,10 +452,71 @@ TADA_CheckRequiredFields <- function(.data) {
 #' reducedcols_Data_Nutrients_UT <- TADA_RetainRequired(Data_Nutrients_UT)
 #'
 TADA_RetainRequired <- function(.data) {
```
This needs review. I combined functions TADA_RetainRequired and TADA_AutoFilter here. Does that make sense or are they may be better left as separate functions? Possible new name suggestion: maybe TADA_ReduceCols

cristinamullin commented on R/Transformations.R:
```diff
@@ -307,13 +333,22 @@ TADA_CalculateTotalNP <- function(.data, sum_ref, daily_agg = c("max", "min", "m
     "TADA.ComparableDataIdentifier",
-    # "TADA.ResultMeasure.MeasureUnitCode",
+    # "TADA.ResultMeasure.MeasureUnitCode", # this may not need to be included since we have TADA.ComparableDataIdentifier
+    # consider adding columns below... 
```
@wokenny13 FYI - these column additions were meant to address this issue with metadata being carried over. Maybe it can be addressed elsewhere in this function (or might not be possible). Needs more research: https://github.com/USEPA/EPATADA/issues/588

cristinamullin committed:
add curl as dependency

cristinamullin committed:
update TADA_CalculateTotalNP

add dataRetrieval depends to EPATADA depends, improve print messages, add example to TADA_CalculateTotalNP, address QC and non-numeric results within TADA_CalculateTotalNP, add TADA_CalculateTotalNP tests

cristinamullin committed:
organize functions

move functions to more applicable .R files

cristinamullin committed:
update docs

cristinamullin committed:
Update Transformations.R

cristinamullin committed:
update TADA_AggregateMeasurements and TADA_RunKeyFlagFunctions

the default should be clean = FALSE for TADA_AggregateMeasurements and TADA_RunKeyFlagFunctions, add better messages for flag functions

cristinamullin committed:
update TADA_CalculateTotalNP

add more TN/TP combinations, update TADA_CalculateTotalNP documentation, add tests

cristinamullin committed:
update TADA_CalculateTotalNP

add additional matches, handle aggregated values, handle non-numeric results

cristinamullin committed:
fix NA bug in TADA_CalculateTotalNP

cristinamullin committed:
remove duplicated info, improve TADA_CalculateTotalNP flags

cristinamullin committed:
fix TADA_CalculateTotalNP and TADA_IDCensoredData logic and create tests

fix tests and add more, update logic

cristinamullin commented:
ideally, these could be run in any order... or we should think about running censored data handling in autoclean to ensure the proper order....

Suggest to add logic for NA TADA.ResultMeasure.MeasureUnitCode's to TADA_ConvertSpecialChars in future PR

```r

test_that("TADA_SimpleCensoredMethods does not introduce duplicates or NAs in result or unit cols", {
  
  testdat <- TADA_RandomTestingData()
  
  testdat <- TADA_ConvertSpecialChars(testdat, 
                                      col = "TADA.ResultMeasureValue",
                                      clean = TRUE)
  
  # Test to ensure the column is entirely numeric
  expect_true(is.numeric(testdat$TADA.ResultMeasureValue))
  
  # Test to ensure value column does not contain any NA values
  expect_true(!any(is.na(testdat$TADA.ResultMeasureValue)))
  
  # TADA_ConvertSpecialChars does not handle this yet 8/11/25
  # # Test to ensure unit column does not contain any NA values
  # expect_true(!any(is.na(testdat$TADA.ResultMeasure.MeasureUnitCode)))
  
  testdat2 <- TADA_SimpleCensoredMethods(testdat,
                                         nd_method = "multiplier",
                                         nd_multiplier = 0.5,
                                         od_method = "as-is",
                                         od_multiplier = "null")
  
  testdat3 <- TADA_ConvertSpecialChars(testdat3, 
                                      col = "TADA.ResultMeasureValue",
                                      clean = TRUE)
  
  # Test to ensure the column is entirely numeric
  expect_true(is.numeric(testdat3$TADA.ResultMeasureValue))
  
  # Test to ensure value column does not contain any NA values
  expect_true(!any(is.na(testdat3$TADA.ResultMeasureValue)))
  
  # # Test to ensure unit column does not contain any NA values
  # expect_true(!any(is.na(testdat2$TADA.ResultMeasure.MeasureUnitCode)))
})
```

cristinamullin committed:
update tests
